piet-direct2d/src/lib.rs

```diff
@@ -555,20 +559,358 @@ impl<'a> TextLayoutBuilder for D2DTextLayoutBuilder<'a> {
+
+        let layout = text_layout.new_text_layout(&font, "piet").build().unwrap();
+        let piet_width = layout.width();
+
+        let layout = text_layout.new_text_layout(&font, "pie").build().unwrap();
+        let pie_width = layout.width();
+
+        let layout = text_layout.new_text_layout(&font, "pi").build().unwrap();
+        let pi_width = layout.width();
+
+        let layout = text_layout.new_text_layout(&font, "p").build().unwrap();
+        let p_width = layout.width();
+
+        let layout = text_layout.new_text_layout(&font, "").build().unwrap();
+        let null_width = layout.width();
+
+        let full_layout = text_layout.new_text_layout(&font, "piet text!").build().unwrap();
+
+        assert_eq!(full_layout.hit_test_text_position(3, true).map(|p| p.point.x as f64), Some(piet_width));
+        assert_eq!(full_layout.hit_test_text_position(2, true).map(|p| p.point.x as f64), Some(pie_width));
+        assert_eq!(full_layout.hit_test_text_position(1, true).map(|p| p.point.x as f64), Some(pi_width));
+        assert_eq!(full_layout.hit_test_text_position(0, true).map(|p| p.point.x as f64), Some(p_width));
+
+        assert_eq!(full_layout.hit_test_text_position(0, false).map(|p| p.point.x as f64), Some(null_width));
+        assert_eq!(full_layout.hit_test_text_position(9, true).map(|p| p.point.x as f64), Some(full_layout.width()));
+        assert_eq!(full_layout.hit_test_text_position(10, true).map(|p| p.point.x as f64), None);
+    }
+
+    #[test]
+    fn test_hit_test_text_position_complex_0() {
+        let dwrite = directwrite::factory::Factory::new().unwrap();
+
+        let input = "é";
+
+        let mut text_layout = D2DText::new(&dwrite);
+        let font = text_layout.new_font_by_name("Segoe UI", 12.0).build().unwrap();
+        let layout = text_layout.new_text_layout(&font, input).build().unwrap();
+
+        assert_eq!(layout.hit_test_text_position(0, true).map(|p| p.point.x), Some(layout.width()));
+        assert_eq!(input.len(), 2);
+
+        // unicode segmentation is wrong on this one for now.
+        //let input = "🤦\u{1f3fc}\u{200d}\u{2642}\u{fe0f}";
+
+        //let mut text_layout = D2DText::new();
+        //let font = text_layout.new_font_by_name("Segoe UI", 12.0).build().unwrap();
+        //let layout = text_layout.new_text_layout(&font, input).build().unwrap();
+
+        //assert_eq!(input.graphemes(true).count(), 1);
+        //assert_eq!(layout.hit_test_text_position(0, true).map(|p| p.point_x as f64), Some(layout.width()));
+        //assert_eq!(input.len(), 17);
+
+        let input = "\u{0023}\u{FE0F}\u{20E3}"; // #️⃣
+
+        let mut text_layout = D2DText::new(&dwrite);
+        let font = text_layout.new_font_by_name("Segoe UI", 12.0).build().unwrap();
+        let layout = text_layout.new_text_layout(&font, input).build().unwrap();
+
+        assert_eq!(layout.hit_test_text_position(0, true).map(|p| p.point.x), Some(layout.width()));
+        assert_eq!(input.len(), 7);
```

Comment:
For style, I think these would make sense right after the def of `input` and before the querying - they seem to be sanity-checks rather than something that actually tests the lib.

Reply:
agreed